package_managers/bootstrap_image.bzl

```diff
@@ -0,0 +1,164 @@
+        content = build_contents,
+        is_executable = True,
+    )
+    
+    ctx.actions.run(
+        outputs = [ctx.outputs.packages_tar],
+        inputs = ctx.attr.download_pkgs.default_runfiles.files.to_list() +
+                 [get_status,
+                  ctx.file.image_tar, fetch_or_download, ],
+        executable = fetch_or_download,
+        mnemonic = "RunFetchOrDownload",
+        use_default_shell_env = True,
+    )
+    # this is not executed when you call the install_pkgs rule. 
+    # Only gets executed when you run _fetch target
+    put_status = git_store_put(
+        ctx = ctx,
+        store_location = ctx.attr.store_location,
+        artifact = ctx.outputs.packages_tar,
+        key = store_key
+    )
+
+    return struct(
+        files = depset([ctx.outputs.packages_tar,]),
+        runfiles = ctx.runfiles(files = ctx.attr.download_pkgs.default_runfiles.files.to_list() +
+                                        [put_status, 
+                                         ctx.file.image_tar])
+    )
+
+fetch_or_download_pkgs = rule(
+    attrs = dict({
+        "image_tar": attr.label(
+            default = Label("//ubuntu:ubuntu_16_0_4_vanilla.tar"),
+            allow_files = True,
+            single_file = True,
+        ),
+       "download_pkgs": attr.label(
+            cfg = "target",
+            executable = True,
+            allow_files = True,
+        ),
+       "store_location": attr.string(),
+       "date": attr.string(), 
+       "output_image_name": attr.string(),
+    }.items() + _git_store_tools.items()),
+    outputs = {
+      "packages_tar": "%{name}.tar",
+    },
+    implementation = _impl,
+)
+
+
+
+"""Bootstrap images with packages from package manager or given location.
+
+This rule builds an image with packages either by downloading packages from
+package manager or given location.
+
+Args:
+  name: A unique name for this rule.
+  image_tar: The image tar for the container used to download packages.
+  package_manager_genrator: A target which generates a script using
+       package management tool e.g apt-get, dpkg to downloads packages.
+  store: A target to store the downloaded packages or retrieve previously downloaded packages.
+  additional_repos: list of additional debian package repos to use, in sources.list format
+"""
+
+def bootstrap_image_macro(name, image_tar, packages, store_location, date, output_image_name, additional_repos=[]):
+  """Downloads packages within a container
+  This rule creates a script to download packages within a container.
+  The script bunldes all the packages in a tarball.
+  Args:
+    name: A unique name for this rule.
+    image_tar: The image tar for the container used to download packages.
+    packages: list of packages to download. e.g. ['curl', 'netbase']
+    additional_repos: list of additional debian package repos to use, in sources.list format
+  """
+
+  download_pkgs(
+       name = "{0}_download".format(name),
```

Comment:
It might be worth saving this as a variable to make it clear that it's used below in fetch_or_download.

Reply:
done